In [6]:
%load_ext sql

In [4]:
import json
from sqlalchemy import create_engine
with open("../../db_config.local.json", encoding="utf-8") as f:
    cfg = json.load(f)
engine = create_engine(cfg["connection_string"] + "libdemo")

In [7]:
%sql engine

In [8]:
%sql SHOW TABLES

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/libdemo'

1 rows affected.

Tables_in_libdemo
borrow_records


In [9]:
import pandas as pd

bor = pd.read_csv("data/borrow_records.csv", encoding="utf-8-sig")
rdr = pd.read_csv("data/readers.csv", encoding="utf-8-sig").drop_duplicates()
rdr["学院"] = rdr["学院"].str.strip()  # 先清洗再合并——脏键会悄悄丢掉匹配

m = bor.merge(rdr, on="读者编号", suffixes=("_借阅", "_读者"))

In [11]:
rdr.to_sql("reader_archive", con=engine, if_exists="replace", index=False)

148

In [12]:
%%sql
SELECT COUNT(*) FROM reader_archive r
LEFT JOIN borrow_records b ON r.读者编号 = b.读者编号
WHERE b.读者编号 IS NULL;   -- 右表配不上的 = 从未借过书的人

Running query in 'mysql+pymysql://root:***@127.0.0.1:3306/libdemo'

1 rows affected.

COUNT(*)
2


In [13]:
raw = pd.read_csv("data/readers.csv", encoding="utf-8-sig")
print("原始学院取值:", raw["学院"].unique().tolist())  # 注意 '历史 ' 尾部有空格
print("去重后行数:", len(rdr))
print("年级缺失:", rdr["年级"].isna().sum())

orphan = bor[~bor["读者编号"].isin(rdr["读者编号"])]
print("孤儿记录:", len(orphan), "条 ->", orphan["读者编号"].tolist())

原始学院取值: ['历史', '文学', '信息管理', '法学', '计算机', '历史 ']
去重后行数: 148
年级缺失: 2
孤儿记录: 2 条 -> ['R1148', 'R1149']


In [14]:
print("合并后行数:", len(m))

per = m.groupby("学院_读者").agg(总量=("借阅量", "sum"), 读者数=("读者编号", "nunique"))
per["人均"] = (per["总量"] / per["读者数"]).round(2)
per.sort_values("人均", ascending=False)

合并后行数: 146


,总量,读者数,人均
学院_读者,,,
计算机,13233.0,31,426.87
信息管理,3189.0,26,122.65
历史,2877.0,28,102.75
法学,2984.0,30,99.47
文学,3081.0,31,99.39


In [15]:
m.to_csv("data/merged.csv", index=False, encoding="utf-8-sig")
print("merged.csv 已生成:", len(m), "行")

merged.csv 已生成: 146 行


In [16]:
n_bad = (m["学院_借阅"] != m["学院_读者"]).sum()
print("学院不一致行数:", n_bad, "/", len(m))

m2 = m[m["借阅量"] != 9999]  # 去掉 9999 离群值
per2 = m2.groupby("学院_读者").agg(
    总量=("借阅量", "sum"), 读者数=("读者编号", "nunique")
)
per2["人均"] = (per2["总量"] / per2["读者数"]).round(2)
per2.sort_values("人均", ascending=False)

学院不一致行数: 116 / 146


,总量,读者数,人均
学院_读者,,,
信息管理,3189.0,26,122.65
计算机,3234.0,30,107.80
历史,2877.0,28,102.75
法学,2984.0,30,99.47
文学,3081.0,31,99.39


# 结论汇总
题1：对。merge 默认 how="inner"，只保留两边都有的键，2 条孤儿记录被自动丢弃（148→146 已验证）。
题8：决定**保留 + 标记"待核"**。R1148/R1149 紧挨档案表最后编号 R1147，最像办证未建档或证件号录错；借阅流水是出纳凭证，删了就永久丢线索。做法：借阅表加"档案状态=待核"列，人工核对后补建档案。
题9：学院不一致 116/146 行（约 80%）——两套系统从未对账；更相信读者档案表（登记时录入的权威数据；借阅流水是办理时快速录入）。9999 是离群值，删掉后人均第一名从"计算机 426.87"变成"信息管理 122.65"——离群值第三次改变排行榜结论。
